pkg/gatewayserver/io/basicstationlns/basicstationlns.go

```diff
@@ -0,0 +1,362 @@
+	}
+	data, err = json.Marshal(res)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to marshal response message")
+		writeDiscoverError(ctx, ws, "Router not provisioned")
+		return err
+	}
+	if err := ws.WriteMessage(websocket.TextMessage, data); err != nil {
+		logger.WithError(err).Warn("Failed to write discover response message")
+		return err
+	}
+	logger.Debug("Sent discover response message")
+	return nil
+}
+
+func (s *srv) handleTraffic(c echo.Context) error {
+	var latestUpstreamXTime int64
+	id := c.Param("id")
+	var auth string
+	auth = c.Request().Header.Get(echo.HeaderAuthorization)
+	ctx := c.Request().Context()
+	if auth != "" {
+		if !strings.Contains(auth, "Bearer") {
+			auth = fmt.Sprintf("Bearer %s", auth)
+		}
+		md := metadata.New(map[string]string{
+			"id":            id,
+			"authorization": auth,
+		})
+		if ctxMd, ok := metadata.FromIncomingContext(s.ctx); ok {
+			md = metadata.Join(ctxMd, md)
+		}
+		ctx = metadata.NewIncomingContext(s.ctx, md)
+	}
+
+	logger := log.FromContext(ctx).WithFields(log.Fields(
+		"endpoint", "traffic",
+		"remote_addr", c.Request().RemoteAddr,
+	))
+
+	ctx, ids, err := s.server.FillGatewayContext(ctx, ttnpb.GatewayIdentifiers{GatewayID: id})
+	if err != nil {
+		return err
+	}
+
+	// For gateways with valid EUIs and no auth, we provide the link rights ourselves as in the udp frontend.
+	if auth == "" {
+		ctx = rights.NewContext(ctx, rights.Rights{
+			GatewayRights: map[string]*ttnpb.Rights{
+				id: {
+					Rights: []ttnpb.Right{ttnpb.RIGHT_GATEWAY_LINK},
+				},
+			},
+		})
+	}
+
+	uid := unique.ID(ctx, ids)
+	ctx = log.NewContextWithField(ctx, "gateway_uid", uid)
+
+	conn, err := s.server.Connect(ctx, s, ids)
+	if err != nil {
+		logger.WithError(err).Warn("Failed to connect")
+		return err
+	}
+	if err := s.server.ClaimDownlink(ctx, ids); err != nil {
+		logger.WithError(err).Error("Failed to claim downlink")
+		return err
+	}
+	defer func() {
+		if err := s.server.UnclaimDownlink(ctx, ids); err != nil {
+			logger.WithError(err).Error("Failed to unclaim downlink")
+		}
+	}()
+
+	ws, err := s.upgrader.Upgrade(c.Response(), c.Request(), nil)
+	if err != nil {
+		logger.WithError(err).Debug("Failed to upgrade request to websocket connection")
+		return err
+	}
+	defer ws.Close()
+
+	fp := conn.FrequencyPlan()
+
+	go func() {
+		for {
+			select {
+			case <-conn.Context().Done():
+				return
+			case down := <-conn.Down():
+				dlTime := time.Now()
+				scheduledMsg := down.GetScheduled()
+				latestXTime := atomic.LoadInt64(&latestUpstreamXTime)
+
+				// Convert to microseconds
+				concentratorTimeMs := conn.GetConcentratorTime(scheduledMsg.Timestamp) / 1000
+
+				// The first 16 bits of XTime gets the session ID from the upstream latestXTime and the other 48 bits are concentrator timestamp accounted for rollover.
+				xTime := int64(uint64(latestXTime)&0xFFFF000000000000 | uint64(concentratorTimeMs)&0x0000FFFFFFFFFFFF)
```

Comment:
Make this as easy and clear as possible;

```go
xTime := int64(sessionID) << 48 | conn.GetConcentratorTime(scheduledMsg.Timestamp) / int64(time.Microsecond)
```

Reply:
Yeah alright. the reason for keeping the existing int64 is to continue using the `atomic` RW. Updated to use a regular mutex.

Reply:
But do we need a mutex? Is this value going to change during a connection, or is it just a value that needs to become available through uplink?

Reply:
Refs https://github.com/TheThingsNetwork/lorawan-stack/pull/762#discussion_r301114566

Reply:
According to the code, it is (currently) constant per connection. But there are 2 reasons why it's better to keep updating this value on every uplink:
1. Constant session ID per connection is [current](https://github.com/lorabasics/basicstation/blob/c29b8502f8c715daecec6666835da6e981dc820a/src/timesync.c#L379) behavior. The fact that the example python bridge [updates this value every uplink](https://github.com/lorabasics/basicstation/blob/c29b8502f8c715daecec6666835da6e981dc820a/examples/station2pkfwd/router.py#L135) suggests, that this may change in the future.
2. Even if we update the variable `sessionID` only once per connection, it might be read in the DL routine concurrently while this is happening. Even if it's rare, I still think it's better to protect the concurrent read in this case.